web/src/App.test.js

```diff
@@ -1,9 +1,52 @@
 import { screen } from "@testing-library/react";
-import { render } from "./test-utils";
+import userEvent from "@testing-library/user-event";
+import { authRender } from "./test-utils";
 import App from "./App";
+import InstallerClient from "./lib/InstallerClient";
 
-test("renders the App", () => {
-  render(<App />);
-  const title = screen.getByText(/Welcome to D-Installer/i);
-  expect(title).toBeInTheDocument();
+jest.mock("./lib/InstallerClient");
+jest.mock("./Installer", () => {
+  return {
+    __esModule: true,
+    default: () => {
+      return <div>Installer Component</div>;
+    }
+  };
+});
+
+describe("when the user is already logged in", () => {
+  beforeEach(() => {
+    InstallerClient.mockImplementation(() => {
+      return {
+        authorize: (_username, _password) => Promise.resolve(false),
+        isLoggedIn: () => Promise.resolve(true),
+        currentUser: () => Promise.resolve("jane")
+      };
+    });
+  });
+
+  it("shows the installer", async () => {
+    authRender(<App />);
+    await screen.findByText("Installer Component");
+  });
+});
+
+describe("when username and password are wrong", () => {
+  beforeEach(() => {
+    InstallerClient.mockImplementation(() => {
+      return {
+        authorize: () => Promise.reject("password does not match"),
+        isLoggedIn: () => Promise.resolve(false),
+        onSignal: jest.fn()
+      };
+    });
+  });
+
+  it("shows an error", async () => {
+    authRender(<App />);
+    userEvent.type(screen.getByLabelText(/Username/i), "john");
+    userEvent.type(screen.getByLabelText(/Password/i), "something");
+    userEvent.click(screen.getByRole("button", { name: /Login/ }));
+    await screen.findByText(/Authentication failed/i);
```

Comment:
this awaits is not so clear for me. Does it mean that in failed case it will wait timeout and then make test failed with error message that given text is not found?

Reply:
Yes, that's it. The react-testing-library documentation contains [an interesting comparison](https://testing-library.com/docs/dom-testing-library/cheatsheet/#queries) of `findBy`, `getBy` and `queryBy`. You can find more information about [findBy](https://testing-library.com/docs/dom-testing-library/api-async#findby-queries) queries.